### PR TITLE
Feat/account recovery improvements

### PR DESCRIPTION
**Pull request by TheLeoB**
## Description (optional)
Add a concise explanation of the changes.

## Testing Steps (optional)
Detail the steps required for the reviewer(s) to verify and test these changes.

### CODE REVIEW

File: packages/blockchain-wallet-v4-frontend/src/data/auth/sagas.ts [hunk 1]
```diff
@@ -373,16 +373,13 @@ export default ({ api, coreSagas, networks }) => {
...
-          if (product === ProductAuthOptions.WALLET) {
-            yield put(actions.router.push('/home'))
+          const verifiedTwoFa = (yield select(
+            selectors.signup.getRecoveryTwoFAVerification
+          )).getOrElse(false)
+          if (!verifiedTwoFa) {
+            yield put(actions.router.push('/setup-two-factor'))
           } else {
-            yield put(actions.router.push('/select-product'))
+            yield put(actions.router.push('/home'))
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/data/auth/sagas.ts:382]: route to select-product instead?

File: packages/blockchain-wallet-v4-frontend/src/data/auth/sagas.ts [hunk 2]
```diff
@@ -392,7 +389,11 @@ export default ({ api, coreSagas, networks }) => {
       }
       yield call(fetchBalances)
       yield call(saveGoals, firstLogin)
-      yield put(actions.goals.runGoals())
+      // We run goals in accountResetSaga in this case
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/data/auth/sagas.ts:392]: expand comment - if it's account reset, we need to wait until user's metadata is written until we run the goals saga, otherwise it will run before /users/current returns. That call let's us know if we should show the 'Resubmit KYC' docs modal

File: packages/blockchain-wallet-v4-frontend/src/data/session/types.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 export type SessionStateType = {
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/data/session/types.ts:1]: adding a new session token entry

File: packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts [hunk 4]
```diff
@@ -314,12 +400,128 @@ export default ({ api, coreSagas, networks }) => {
...
+    if (n === 0) {
+      yield put(actions.form.change(RECOVER_FORM, 'step', RecoverSteps.FORGOT_PASSWORD_EMAIL))
+      yield put(actions.alerts.displayInfo(C.VERIFY_DEVICE_EXPIRY, undefined, true))
+
+      return false
+    }
+    try {
+      yield delay(2000)
+      const response = yield call(api.pollForResetApprovalStatus, sessionToken)
+      // const { recoveryEmail } = yield select(selectors.form.getFormValues(RECOVER_FORM))
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts:413]: leftover mock responses, removed in next commit

File: packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts [hunk 5]
```diff
@@ -314,12 +400,128 @@ export default ({ api, coreSagas, networks }) => {
...
+      yield put(actions.signup.setAccountRecoveryMagicLinkDataEncoded(accountRecoveryDataEncoded))
+      const accountRecoveryData = JSON.parse(
+        base64url.decode(accountRecoveryDataEncoded)
+      ) as AccountRecoveryMagicLinkData
+      const { email, recovery_token: token, userId } = accountRecoveryData
+      const sessionToken = yield select(selectors.session.getRecoverSessionId, email)
+      yield call(api.approveAccountReset, email, sessionToken, token, userId)
+      yield put(actions.signup.accountRecoveryVerifySuccess(true))
+    } catch (e) {
+      // TODO: handle error
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts:481]: add error screen to recovery approval landing page

File: packages/blockchain-wallet-v4-frontend/src/modals/index.tsx [hunk 6]
```diff
@@ -265,6 +268,9 @@ const Modals = (props: Props) => {
         {props.modals.find((modal) => modal.type === ModalName.SEND_CRYPTO_MODAL) ? (
           <SendCrypto />
         ) : null}
+        {props.modals.find((modal) => modal.type === ModalName.SKIP_TWOFA_CONFIRMATION_WARNING) ? (
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/modals/index.tsx:271]: can probably remove this null check

File: packages/blockchain-wallet-v4-frontend/src/scenes/Login/Wallet/EnterPassword/index.tsx [hunk 7]
```diff
@@ -188,46 +187,44 @@ const EnterPasswordWallet = (props: Props) => {
             />
           </CenteredColumn>
         </WrapperWithPadding>
-        <SignupLink platform={magicLinkData?.platform_type} />
+        {/* <SignupLink platform={magicLinkData?.platform_type} /> */}
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/Login/Wallet/EnterPassword/index.tsx:190]: remove this component

File: packages/blockchain-wallet-v4-frontend/src/scenes/Login/Wallet/EnterPassword/index.tsx [hunk 8]
```diff
@@ -188,46 +187,44 @@ const EnterPasswordWallet = (props: Props) => {
...
-              >
-                <FormattedMessage
-                  id='scenes.login.wallet.mobile_login.notification'
-                  defaultMessage='Ensure that notifications are enabled for the Blockchain.com app.'
-                />
-              </Text>
-            </TextColumn>
-          </MobileAuthSideWrapper>
-        </SideWrapper>
+        <MobileAuthWrapper>
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/Login/Wallet/EnterPassword/index.tsx:193]: moves mobile QR code component underneath PW modal

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/CheckInbox/index.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        setSentState('checkEmail')
+      }, 5000)
+    }
+  }, [disabled, sentState])
+
+  const hasErrorAlert = props.alerts.find((alert) => alert.message === RECOVERY_EMAIL_SENT_ERROR)
+
+  const handleSubmit = (e) => {
+    e.preventDefault()
+    // trigger email request for recovery token
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/CheckInbox/index.tsx:38]: no longer need this

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/ResetAccount/NewPassword/index.tsx [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    e.preventDefault()
+    const { formValues, language, signupActions } = props
+    if (accountRecoveryV2Flag) {
+      signupActions.resetAccountV2({
+        email: formValues?.recoveryEmail,
+        language,
+        password: formValues.resetAccountPassword
+      })
+    } else {
+      // legacy account recovery for FF
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/ResetAccount/NewPassword/index.tsx:46]: change comment for feature flag

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/ResetAccount/TwoFAConfirmation/index.tsx [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+  const handleSubmit = (e) => {
+    e.preventDefault()
+    signupActions.verifyTwoFaForRecovery({
+      code: formValues?.twoFACode,
+      email: formValues.recoveryEmail
+    })
+  }
+
+  //   const accountLocked =
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/ResetAccount/TwoFAConfirmation/index.tsx:53]: can likely remove these, have a todo to make sure i'm handling incorrect 2fa correctly

File: packages/blockchain-wallet-v4-frontend/src/scenes/TwoStepVerification/Authenticator/template.qrcode.tsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+          <FormattedMessage
+            id='scenes.login.upgrade.googleAuth.text'
+            defaultMessage='With your Google Authenticator app, scan the QR code below to make a secure connection.'
+          />
+        </Text>
+        {props.googleAuthSecretUrl && (
+          <QRCodeWrapper size={150} value={props.googleAuthSecretUrl} />
+        )}
+        {/*
+        <CopyClipboard address={props.data.secret || ''} /> */}
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/TwoStepVerification/Authenticator/template.qrcode.tsx:49]: will probably need this for mobile view? todo

File: packages/blockchain-wallet-v4-frontend/src/scenes/TwoStepVerification/ChoooseTwoFA/index.tsx [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+            size='16px'
+            weight={600}
+            data-e2e='skipTwoFA'
+            style={{ cursor: 'pointer' }}
+          >
+            <FormattedMessage id='modals.wallet.welcome.sb.skip' defaultMessage='Skip Setup' />
+          </SkipLinkText>
+        </LinkContainer>
+      </SkipLink>
+      {/* <ButtonPanel onClick={() => props.setFormStep(TwoFactorSetupSteps.SMS_SETUP)}>
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/TwoStepVerification/ChoooseTwoFA/index.tsx:163]: will add comment, we want to move away from SMS 2fa so removed it from here

File: packages/blockchain-wallet-v4-frontend/src/scenes/TwoStepVerification/SMS/template.setup.tsx [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+                defaultMessage='Change Mobile Number'
+              />
+            </Text>
+          </CenteredColumn>
+          <Button
+            nature='primary'
+            data-e2e='submitCode'
+            fullwidth
+            height='48px'
+            // TODO: Add real on click event
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/TwoStepVerification/SMS/template.setup.tsx:78]: may not be necc since we're deprecating sms

File: packages/blockchain-info-components/src/Images/Images.ts [hunk 15]
```diff
@@ -329,7 +330,8 @@ const Images = {
   'world-alert2': worldAlert2,
   'world-alert3': worldAlert3,
   'yodlee-connect': yodleeConnect,
-  yubikey
+  yubikey,
```
Comment (TheLeoB) [inline packages/blockchain-info-components/src/Images/Images.ts:333]: has been added to constellation

File: packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts [hunk 16]
```diff
@@ -228,6 +248,72 @@ export default ({ api, coreSagas, networks }) => {
...
+          exchange_lifetime_token: mercuryLifetimeToken,
+          exchange_user_id: userCredentialsId,
+          nabu_lifetime_token: token,
+          nabu_user_id: userId
+        })
+      )
+
+      sessionStorage.removeItem('accountRecovery')
+
+      // if user is resetting their account and
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts:288]: remove this

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/EmailAuthLanding/template.loading.tsx [hunk 17]
```diff
@@ -0,0 +1,36 @@
+import React from 'react'
+import { FormattedMessage } from 'react-intl'
+import styled from 'styled-components'
+
+import { SpinningLoader, Text } from 'blockchain-info-components'
+
+const Wrapper = styled.div`
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/EmailAuthLanding/template.loading.tsx:7]: use flex from constellation

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/EmailAuthLanding/template.success.tsx [hunk 18]
```diff
@@ -0,0 +1,37 @@
+import React from 'react'
+import { FormattedMessage } from 'react-intl'
+import styled from 'styled-components'
+
+import { Icon, Text } from 'blockchain-info-components'
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/EmailAuthLanding/template.success.tsx:5]: use icon from constellation

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/RecoveryOptions/index.tsx [hunk 19]
```diff
@@ -28,27 +28,32 @@ const FormBody = styled.div`
 const IconTextRow = styled.div`
   display: flex;
```
Comment (TheLeoB) [inline packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/RecoveryOptions/index.tsx:29]: use contellation

File: package.json [hunk 20]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "blockchain-wallet-v4",
-  "version": "4.78.4",
+  "version": "4.77.2-recovery-test",
```
Comment (blockdylanb) [inline package.json:3]: this will probably conflict when you try to merge, you may just want to undo this change unless you def need it.

File: packages/blockchain-wallet-v4-frontend/src/data/auth/sagas.ts (hunk 1 above)
Comment (blockdylanb) [reply packages/blockchain-wallet-v4-frontend/src/data/auth/sagas.ts:382]: still an issue?

File: packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts [hunk 21]
```diff
@@ -314,12 +387,108 @@ export default ({ api, coreSagas, networks }) => {
...
+    if (n === 0) {
+      yield put(actions.form.change(RECOVER_FORM, 'step', RecoverSteps.FORGOT_PASSWORD_EMAIL))
+      yield put(actions.alerts.displayInfo(C.VERIFY_DEVICE_EXPIRY, undefined, true))
+
+      return false
+    }
+    try {
+      yield delay(2000)
+      const response = yield call(api.pollForResetApprovalStatus, sessionToken)
+      // if (response?.status === 'true') {
```
Comment (blockdylanb) [inline packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts:400]: need this or can it be deleted?

File: packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts [hunk 22]
```diff
@@ -314,12 +387,108 @@ export default ({ api, coreSagas, networks }) => {
...
+      }
+      if (response?.status === AccountRecoveryApprovalStatusType.INVALID) {
+        yield put(actions.form.change(RECOVER_FORM, 'step', RecoverSteps.FORGOT_PASSWORD_EMAIL))
+        yield put(actions.alerts.displayError(C.VERIFY_DEVICE_FAILED, undefined, true))
+        return false
+      }
+    } catch (error) {
+      return false
+    }
+    return yield call(pollForResetApproval, sessionToken, n - 1)
```
Comment (blockdylanb) [inline packages/blockchain-wallet-v4-frontend/src/data/signup/sagas.ts:414]: You might think about refactoring this to the redux-saga `retry` effect. https://redux-saga.js.org/docs/api/#retrymaxtries-delay-fn-args It might not be exactly what you're looking for though so just a suggestion.

File: packages/blockchain-wallet-v4-frontend/src/data/signup/slice.ts [hunk 23]
```diff
@@ -30,7 +35,19 @@ const signupSlice = createSlice({
   initialState,
   name: 'signup',
   reducers: {
+    accountRecoveryVerify: (state, action?) => {},
+    accountRecoveryVerifyFailure: (state, action) => {
```
Comment (blockdylanb) [inline packages/blockchain-wallet-v4-frontend/src/data/signup/slice.ts:39]: is it possible to pass the type of action.payload? `(state, action: PayloadAction<AcountRecoveryVerifyPayload>)`

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/EmailAuthLanding/template.loading.tsx [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+
+const LoadingWrapper = styled.div`
+  position: relative;
+  display: flex;
+  justify-content: center;
+`
+
+const Loading = () => {
+  return (
+    <Flex alignItems='center' flexDirection='column'>
```
Comment (blockdylanb) [inline packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/EmailAuthLanding/template.loading.tsx:16]: 👏🏻

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/RecoveryPhrase/index.tsx [hunk 25]
```diff
@@ -9,27 +8,21 @@ import { RECOVER_FORM } from '../model'
...
-  InjectedFormProps<{}, Props> & Props,
-  StateProps
-> {
-  constructor(props) {
-    super(props)
-    this.state = {
-      step: 1
-    }
-  }
+const RecoveryPhraseContainer = (props: Props) => {
```
Comment (blockdylanb) [inline packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/RecoveryPhrase/index.tsx:11]: 👏🏻

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/ResetAccount/NewPassword/index.tsx (hunk 10 above)
Comment (blockdylanb) [reply packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/ResetAccount/NewPassword/index.tsx:46]: still neede?

File: packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/CheckInbox/index.tsx [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+        fullwidth
+        height='48px'
+        data-e2e='loginResendEmail'
+        disabled={disabled && !hasErrorAlert}
+        // @ts-ignore
+        onClick={(e: SyntheticEvent) => {
+          setDisabled(true)
+          setSentState('sent')
+          // TODO: correctly handle this resend email button
+          handleSubmit(e)
```
[the remark this answers is not in the document]
Comment (TheLeoB) [reply packages/blockchain-wallet-v4-frontend/src/scenes/RecoverWallet/CheckInbox/index.tsx:73]: Yes, good call.

**Review by blockdylanb — APPROVED** [review]
